homeassistant/components/lgthinq/__init__.py

```diff
@@ -0,0 +1,331 @@
+        country_code=entry.data.get(CONF_COUNTRY, DEFAULT_COUNTRY),
+        client_id=client_id,
+        access_token=access_token,
+    )
+    device_registry = dr.async_get(hass)
+
+    # Setup and register devices.
+    await async_setup_devices(hass, thinq, device_registry, entry)
+
+    # Set up all platforms for this device/entry.
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    # Set up custom services.
+    async_setup_hass_services(hass, entry)
+
+    # Clean up devices they are no longer in use.
+    async_cleanup_device_registry(
+        entry.runtime_data.device_map.values(), device_registry, entry.entry_id
+    )
+
+    return True
+
+
+async def async_setup_devices(
+    hass: HomeAssistant,
+    thinq: ThinQ,
+    device_registry: dr.DeviceRegistry,
+    entry: ThinqConfigEntry,
+) -> None:
+    """Set up and register devices."""
+    _LOGGER.warning("Starting set up devices")
+
+    entry.runtime_data.device_map.clear()
+
+    # Get a device list from the server.
+    response: ThinQApiResponse = await thinq.async_get_device_list()
+    if not ThinQ.is_success(response):
+        raise ConfigEntryError(
+            response.error_message,
+            translation_domain=DOMAIN,
+            translation_key=response.error_code,
+        )
+
+    device_list = response.body
+    if not device_list or not isinstance(device_list, Collection):
+        return
+
+    # Setup devices.
+    lg_device_list: list[LGDevice] = []
+    task_list = [
+        hass.async_create_task(async_setup_lg_device(hass, thinq, device))
+        for device in device_list
+    ]
+    if task_list:
+        task_result = await asyncio.gather(*task_list)
+        for lg_devices in task_result:
+            if lg_devices:
+                lg_device_list.extend(lg_devices)
+
+    # Register devices.
+    async_register_devices(lg_device_list, device_registry, entry)
+
+
+@callback
+def async_register_devices(
+    lg_device_list: list[LGDevice],
+    device_registry: dr.DeviceRegistry,
+    entry: ThinqConfigEntry,
+) -> None:
+    """Register devices to the device registry."""
+    if lg_device_list:
+        for lg_device in lg_device_list:
+            device_entry = device_registry.async_get_or_create(
+                config_entry_id=entry.entry_id,
+                **lg_device.device_info,
+            )
+            _LOGGER.debug(
+                "Register device: device_id=%s, device_entry_id=%s",
+                lg_device.id,
+                device_entry.id,
+            )
+            entry.runtime_data.device_map[device_entry.id] = lg_device
+
+
+@callback
+def async_cleanup_device_registry(
+    lg_devices: Collection[LGDevice],
+    device_registry: dr.DeviceRegistry,
+    entry_id: str,
+) -> None:
+    """Clean up device registry."""
+    new_device_unique_ids: list[str] = [device.unique_id for device in lg_devices]
+    existing_entries: list[dr.DeviceEntry] = dr.async_entries_for_config_entry(
```

Comment:
This is already typed upstream?

```suggestion
    existing_entries = dr.async_entries_for_config_entry(
```

Reply:
Removed.